c10/core/inference_mode.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+namespace c10 {
+
+struct TORCH_API InferenceMode {
+  static bool is_enabled();
+  static void set_enabled(bool enabled);
+};
+
+// A RAII, thread local (!) guard that enables or disables inference mode upon
+// construction, and sets it back to the original value upon destruction.
+struct TORCH_API AutoInferenceMode {
```

Comment:
Bikeshed: I'd actually vote for naming this simply InferenceMode (and renaming the struct on line 8 to something else, or maybe even just stuffing the static methods on the same class). I don't think you need the "Auto" because "Mode" should already make it clear that an RAII-ey thing is happening here.

At the *very least*, there shouldn't be an attractive nuisance InferenceMode struct in public interface, because that makes it super easy for someone to write:

```
c10::InferenceMode mode;
```

and it would work! It would compile! Your tests would even pass! But you wouldn't actually have gotten the mode.